### PR TITLE
Improve linker friendliness for RouteOptions

### LINKED ISSUE #24723: RouteConstraints aren't linker friendly
Today RouteConstraints are described by a [RouteOptions.ConstraintMap](https://github.com/dotnet/aspnetcore/blob/983e7ed635f4d128edce4191b51869bd7be8aced/src/Http/Routing/src/RouteOptions.cs#L83). The issue with this approach is the linker doesn't have way to know that these types are activated elsewhere in the system. We need an API that lets us describe adding a constraint that is type safe and lets us describe that we want to preserve constructors:

```C#
public class RouteOptions
{
    public void AddParameterPolicy<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]T>(string text) where T : IParameterPolicy;
    public void AddParameterPolicy(string text, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]Type constraintType);
}
```

This would let the linker know that these constraint types should be kept around.

Here's the current warning:

```
C:\dev\git\aspnetcore\src\Http\Routing\src\ParameterPolicyActivator.cs(93,13): Trim analysis warning IL2006: Microsoft.AspNetCore.Routing.ParameterPolicyActivator.CreateParameterPolicy(IServiceProvider,Type,String): The parameter 'parameterPolicyType' of method 'Microsoft.AspNetCore.Routing.ParameterPolicyActivator.CreateParameterPolicy(IServiceProvider,Type,String)' with dynamically accessed member kinds 'None' is passed into the implicit 'this' parameter of method 'System.Type.GetConstructors()' which requires dynamically accessed member kinds 'PublicConstructors'. To fix this add DynamicallyAccessedMembersAttribute to it and specify at least these member kinds 'PublicConstructors'. [C:\Users\davifowl\source\repos\WebApplication60\WebApplication60\WebApplication60.csproj]
```

### PR DESCRIPTION
**Pull request by davidfowl**
- Added API to allows setting parameter policies in a linker friendly way
- Added tests
- Updated solution filter

Fixes #24723

### CODE REVIEW

**Comment by davidfowl** [comment]
Feel free to review the words. I didn't put too much effort into it.

File: src/Http/Routing/src/RouteOptions.cs [hunk 1]
```diff
@@ -124,7 +125,32 @@ internal ICollection<EndpointDataSource> EndpointDataSources
             return defaults;
         }
 
-        // This API could be exposed on RouteOptions
+        /// <summary>
+        /// Adds or overwrites the paramter policy with the associated route pattern token.
```
Comment (javiercn) [inline src/Http/Routing/src/RouteOptions.cs:129]:
```suggestion
        /// Adds or overwrites the parameter policy with the associated route pattern token.
```

**Review by javiercn — APPROVED** [review]
LGTM

File: src/Http/Routing/src/RouteOptions.cs [hunk 2]
```diff
@@ -124,7 +125,32 @@ internal ICollection<EndpointDataSource> EndpointDataSources
...
+        /// </summary>
+        /// <typeparam name="T">The parameter policy type.</typeparam>
+        /// <param name="token">The route token used to apply the parameter policy.</param>
+        public void SetParameterPolicy<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]T>(string token) where T : IParameterPolicy
+        {
+            ConstraintMap[token] = typeof(T);
+        }
+
+        /// <summary>
+        /// Adds or overwrites the paramter policy with the associated route pattern token.
```
Comment (Kahbazi) [inline src/Http/Routing/src/RouteOptions.cs:139]:
```suggestion
        /// Adds or overwrites the parameter policy with the associated route pattern token.
```

File: src/Http/Routing/src/RouteOptions.cs [hunk 3]
```diff
@@ -72,6 +72,7 @@ internal ICollection<EndpointDataSource> EndpointDataSources
         /// </summary>
         public IDictionary<string, Type> ConstraintMap
         {
+            [RequiresUnreferencedCode($"The linker cannot determine what constraints are being added via the ConstraintMap property. Prefer {nameof(RouteOptions)}.{nameof(SetParameterPolicy)} instead.")]
```
Comment (davidfowl) [inline src/Http/Routing/src/RouteOptions.cs:75]: @pranavkm does this sound ok to you?

Comment (davidfowl) [reply src/Http/Routing/src/RouteOptions.cs:75]: Also cc @eerhardt

Comment (pranavkm) [reply src/Http/Routing/src/RouteOptions.cs:75]:
```suggestion
            [RequiresUnreferencedCode($"The linker cannot determine what constraints are being added via the ConstraintMap property. Prefer {nameof(RouteOptions)}.{nameof(SetParameterPolicy)} instead for setting constraints. This warning can be suppressed if this property is being used to read of delete constraints.")]
```

Maybe?

**Review by halter73 — APPROVED** [review]

File: src/Http/Routing/src/RouteOptions.cs [hunk 4]
```diff
@@ -72,6 +72,7 @@ internal ICollection<EndpointDataSource> EndpointDataSources
         /// </summary>
         public IDictionary<string, Type> ConstraintMap
         {
+            [RequiresUnreferencedCode($"The linker cannot determine what constraints are being added via the ConstraintMap property. Prefer {nameof(RouteOptions)}.{nameof(SetParameterPolicy)} instead for setting constraints. This warning can be suppressed if this property is being used to read of delete constraints.")]
```
Comment (eerhardt) [inline src/Http/Routing/src/RouteOptions.cs:75]: Don't say `linker` in public facing docs/strings. Instead, we prefer the term `trimmer` or `trimming`.

> is being used to read of delete constraints.

I'm not sure I understand what that means. Is there a type-o or missing word?

File: src/Http/Routing/src/RouteOptions.cs [hunk 5]
```diff
@@ -72,6 +72,7 @@ internal ICollection<EndpointDataSource> EndpointDataSources
...
             get
             {
                 return _constraintTypeMap;
```
Comment (eerhardt) [inline src/Http/Routing/src/RouteOptions.cs:78]: Don't you also need a `RequiresUnreferencedCode` attribute on the setter?

File: src/Http/Routing/src/RouteOptions.cs (hunk 4 above)
Comment (davidfowl) [reply src/Http/Routing/src/RouteOptions.cs:75]: ah hah

**Pull request merged**